src/java.base/share/classes/jdk/internal/foreign/LayoutRecordMapper.java

```diff
@@ -0,0 +1,641 @@
+                ctor = MethodHandles.filterArguments(ctor, i, handles[i]);
+            }
+
+            var mt = MethodType.methodType(type, MemorySegment.class);
+            // Fold the many identical MemorySegment arguments into a single argument
+            ctor = MethodHandles.permuteArguments(ctor, mt, IntStream.range(0, handles.length)
+                    .map(i -> 0)
+                    .toArray());
+
+            // The constructor MethodHandle is now of type (MemorySegment)T
+            this.ctor = ctor;
+        } catch (IllegalAccessException | NoSuchMethodException e) {
+            throw new IllegalArgumentException("There is no public constructor in '" + type.getName() +
+                    "' for " + Arrays.toString(ctorParameterTypes), e);
+        }
+    }
+
+    private static <U> BinaryOperator<U> throwingMerger() {
+        return (a, b) -> {
+            throw new IllegalArgumentException("Duplicate keys: " + a);
+        };
+    }
+
+    private IllegalArgumentException fail(ComponentAndLayout cl) {
+        throw new IllegalArgumentException(
+                "Unable to map " + cl.layout() + " to " + type.getName() + "." + cl.component().getName());
+    }
+
+    @SuppressWarnings("unchecked")
+    @Override
+    public T apply(MemorySegment segment) {
+        try {
+            return (T) (ctor.invoke(segment));
+        } catch (Throwable e) {
+            throw new IllegalArgumentException(
+                    "Unable to invoke the canonical constructor for " + type.getName() +
+                            " using " + segment, e);
+        }
+    }
+
+    @Override
+    public String toString() {
+        return "LayoutRecordMapper{" +
+                "type=" + type.getName() + ", " +
+                "layout=" + layout + "}";
+    }
+
+    // The parameter is of type ValueLayouts.OfByteImpl and the likes.
+    static Class<? extends ValueLayout> valueLayoutType(ValueLayout vl) {
+        return switch (vl) {
+            case ValueLayout.OfBoolean __ -> ValueLayout.OfBoolean.class;
+            case ValueLayout.OfByte __ -> ValueLayout.OfByte.class;
+            case ValueLayout.OfShort __ -> ValueLayout.OfShort.class;
+            case ValueLayout.OfChar __ -> ValueLayout.OfChar.class;
+            case ValueLayout.OfInt __ -> ValueLayout.OfInt.class;
+            case ValueLayout.OfLong __ -> ValueLayout.OfLong.class;
+            case ValueLayout.OfFloat __ -> ValueLayout.OfFloat.class;
+            case ValueLayout.OfDouble __ -> ValueLayout.OfDouble.class;
+            case AddressLayout __ -> AddressLayout.class;
+        };
+    }
+
+    static MethodHandle findStaticToArray(Class<?> rType,
+                                          Class<?> layoutType,
+                                          Class<?> extra) throws NoSuchMethodException, IllegalAccessException {
+
+        var pTypes = Stream.of(MemorySegment.class, layoutType, long.class, long.class, extra)
```

Comment:
seems a weird way to test if extra is null

Reply:
Using this method for both the primitive and the record case leads to issues. Better use a different way for the recursive record case. Also, you can create a method type from a List::of, which is probably simpler.